adapters/x509_openssl.c

```diff
@@ -162,79 +164,158 @@ static int load_key_RSA(SSL_CTX* ssl_ctx, EVP_PKEY* evp_key)
             result = MU_FAILURE;
         }
         else
         {
-            // Get the Private Key type
-            EVP_PKEY* evp_key = PEM_read_bio_PrivateKey(bio_key, NULL, NULL, NULL);
-            if (evp_key == NULL)
+            // Check the type for the EVP key
+            int evp_type = EVP_PKEY_id(evp_key);
+            if (evp_type == EVP_PKEY_RSA || evp_type == EVP_PKEY_RSA2)
             {
-                log_ERR_get_error("Failure creating private key evp_key");
-                result = MU_FAILURE;
+                if (load_key_RSA(ssl_ctx, evp_key) != 0)
+                {
+                    LogError("failure loading RSA private key cert");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    result = 0;
+                }
             }
             else
             {
-                // Check the type for the EVP key
-                int evp_type = EVP_PKEY_id(evp_key);
-                if (evp_type == EVP_PKEY_RSA || evp_type == EVP_PKEY_RSA2)
+                if (load_ecc_key(ssl_ctx, evp_key) != 0)
                 {
-                    if (load_key_RSA(ssl_ctx, evp_key) != 0)
-                    {
-                        LogError("failure loading RSA private key cert");
-                        result = MU_FAILURE;
-                    }
-                    else
-                    {
-                        result = 0;
-                    }
+                    LogError("failure loading ECC private key cert");
+                    result = MU_FAILURE;
                 }
                 else
                 {
-                    if (load_ecc_key(ssl_ctx, evp_key) != 0)
-                    {
-                        LogError("failure loading ECC private key cert");
-                        result = MU_FAILURE;
-                    }
-                    else
-                    {
-                        result = 0;
-                    }
+                    result = 0;
                 }
+            }
+            EVP_PKEY_free(evp_key);
+        }
+        BIO_free(bio_key);
+    }
+
+    return result;
+}
+
+int x509_openssl_add_engine_key(SSL_CTX* ssl_ctx, const char* x509privatekey_id, ENGINE* engine)
+{
+    int result;
+
+    // Engine functional reference is short-lived, only within this function.
+    if (!ENGINE_init(engine))
+    {
+        log_ERR_get_error("unable to initialize engine.");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (!ENGINE_set_default(engine, ENGINE_METHOD_ALL))
+        {
+            log_ERR_get_error("unable to configure engine.");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            EVP_PKEY* evp_key = ENGINE_load_private_key(engine, x509privatekey_id, NULL, NULL);
 
-                if (result == 0)
+            if (evp_key == NULL)
+            {
+                log_ERR_get_error("failure loading private key from ENGINE");
```

Comment:
nit: Maybe try and unify the log messages here? Something like
`Failure loading private key from ENGINE`
`Unable to configure ENGINE`
`Unable to initialize ENGINE`
`Failure loading private key cert from ENGINE`
etc